src/nearest_neighbors/datasets/heartsteps/loader.py

```diff
@@ -0,0 +1,338 @@
+        return df_steps, df_suggestions
+
+    def _transform_dnn(
+        self,
+        df : Any,
+        users: int = 37,
+        max_study_day: int = 52,
+        day_dec: int = 5,
+        num_measurements: int = 12,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Iteratively transform the processed HeartSteps data into a 4d tensor"""
+        warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
+        final_M = np.zeros((users, max_study_day, day_dec, num_measurements))
+        final_A = np.zeros((users, max_study_day, day_dec))
+
+        for user in range(1, users + 1):
+            for day in range(1, max_study_day + 1):
+                for slot in range(1, day_dec + 1):
+                    try:
+                        df_uds = df.loc[pd.IndexSlice[(user, day, slot)]]
+                        ind = df.index.get_indexer_for(df_uds.index)[0]
+                        df_rng = df.iloc[
+                            np.arange(ind, min(ind + num_measurements, len(df)))
+                        ]
+                        if df.iloc[ind]["avail"]:
+                            # only take send.sedentary as the treatment indicator, could use send.active later
+                            val = df.iloc[ind]["send.sedentary"]
+                            conv_val = val if ~np.isnan(val) else df.iloc[ind]["send"]
+                            final_A[user - 1, day - 1, slot - 1] = int(conv_val)
+                        else:
+                            final_A[user - 1, day - 1, slot - 1] = 2
+
+                        measurements = df_rng["steps"].to_numpy()
+                        if len(measurements) == num_measurements:
+                            final_M[user - 1, day - 1, slot - 1] = measurements
+                        else:
+                            m_pad = np.pad(
+                                measurements,
+                                (0, num_measurements - len(measurements)),
+                                constant_values=np.nan,
+                            )
+                            final_M[user - 1, day - 1, slot - 1] = m_pad
+                    except KeyError:
+
+                        final_A[user - 1, day - 1, slot - 1] = 0
+                        final_M[user - 1, day - 1, slot - 1] = np.full(
+                            num_measurements, np.nan
+                        )
+        final_M = final_M.reshape((users, max_study_day * day_dec, num_measurements))
+        final_A = final_A.reshape((users, max_study_day * day_dec))
+        return final_M, final_A
+
+
+    def _get_mode(self, x: pd.Series) -> pd.Series:
+        if len(pd.Series.mode(x) > 1):
+            return pd.Series.mode(x, dropna=False)[0]
+        else:
+            return pd.Series.mode(x, dropna=False)
+
+
+    def _reind_id(self, df_u: pd.DataFrame) -> pd.DataFrame:
+        """Function to reindex the data to include all time points for each user."""
+        rng = pd.date_range(
```

Comment:
please use better variable names. "RNG" traditionally stands for random number generator even though i see it's clearly a range here